core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -420,396 +375,261 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
-        (wc, refsToUpdateByEmail(wc.email))
-      }
-      val (emailsToAddCatalog, emailsToRemoveCatalog) = emailsFoundWithRefs.partition(update => update._1.catalog)
-
-      dataAccess.workspaceQuery.findWorkspaceUsersAndGroupsWithCatalog(workspaceContext.workspaceId) flatMap { case ((usersWithCatalog, groupsWithCatalog)) =>
-        val usersToAdd = emailsToAddCatalog.collect {
-          case (cat, Left(userref)) if !usersWithCatalog.contains(userref) => (userref, true)
-        }
-        val usersToRemove = emailsToRemoveCatalog.collect {
-          case (cat, Left(userref)) if usersWithCatalog.contains(userref) => (userref, false)
-        }
-        val groupsToAdd = emailsToAddCatalog.collect {
-          case (cat, Right(groupref)) if !groupsWithCatalog.contains(groupref) => (groupref, true)
-        }
-        val groupsToRemove = emailsToRemoveCatalog.collect {
-          case (cat, Right(groupref)) if groupsWithCatalog.contains(groupref) => (groupref, false)
-        }
+    Future.traverse(userPolicyNames) { policyName => samDAO.removeUserFromPolicy(SamResourceTypeNames.workspace, workspaceId, policyName, userEmail.value, userInfo) }.map(_ => ())
+  }
 
-        val emails = emailsNotFound.map { wsCatalog => wsCatalog.email }
-        val users = (usersToAdd ++ usersToRemove).map { case (userRef, catalog) => WorkspaceCatalogResponse(userRef.userSubjectId.value, catalog) }
-        val groups = (groupsToAdd ++ groupsToRemove).map { case (groupRef, catalog) => WorkspaceCatalogResponse(groupRef.groupName.value, catalog) }
+  //TODO: don't construct bucket name here?
+  def addUserToWorkspace(workspaceId: String, existingPolicies: Set[SamPolicyWithNameAndEmail], policyNames: Set[String], userEmail: WorkbenchEmail, inviteUserIfNotFound: Boolean, workspaceName: WorkspaceName): Future[Unit] = {
+    val userPolicyNames = existingPolicies.filter(_.policy.memberEmails.contains(userEmail.value)).map(_.policyName)
 
-        dataAccess.workspaceQuery.insertUserCatalogPermissions(workspaceContext.workspaceId, usersToAdd.map { case (userRef, _) => userRef }) andThen
-          dataAccess.workspaceQuery.deleteUserCatalogPermissions(workspaceContext.workspaceId, usersToRemove.map { case (userRef, _) => userRef }) andThen
-          dataAccess.workspaceQuery.insertGroupCatalogPermissions(workspaceContext.workspaceId, groupsToAdd.map { case (groupRef, _) => groupRef }) andThen
-          dataAccess.workspaceQuery.deleteGroupCatalogPermissions(workspaceContext.workspaceId, groupsToRemove.map { case (groupRef, _) => groupRef }) map { _ =>
-          WorkspaceCatalogUpdateResponseList(users ++ groups, emails)
+    for {
+      //TODO: the removeAll then addAll approach may be a bit too much
+      _ <- Future.traverse(userPolicyNames) { policyName => samDAO.removeUserFromPolicy(SamResourceTypeNames.workspace, workspaceId, policyName, userEmail.value, userInfo) }
```

Comment:
this bit feels dangerous and makes me a little uneasy. we first need to load all of the policies that the user is a member of, remove them from _everything_, and then add them to the policies that they should be a member of. feels like this is a great spot for things to get out of sync from the users expectations